File: built-in plugins dsp/mb-drums.cpp
/**
 * MB Drums
 * Category : instrument
 * Type     : drums
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Punchy drum kit with multiple kits and samples
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DRUMS_H
#define MB_DRUMS_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDrums : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-drums";
    static constexpr const char* PLUGIN_NAME    = "MB Drums";
    static constexpr const char* PLUGIN_TYPE    = "drums";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float punch = 0.7f;  // range [0, 1]
    float tone = 0.5f;  // range [0, 1]
    float room = 0.2f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbDrums() = default;
    ~MbDrums() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.punch = std::clamp(params.punch, 0f, 1f);
        params.tone = std::clamp(params.tone, 0f, 1f);
        params.room = std::clamp(params.room, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Drums
        return input;
    }
};

#endif // MB_DRUMS_H

File: built-in plugins dsp/mb-drums.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbDrumsPlugin: PluginDefinition = {
  id: "mb-drums",
  slug: "mb-drums",
  name: "MB Drums",
  category: "instrument",
  type: "drums",
  version: "1.0.0",
  description: "Punchy drum kit with multiple kits and samples",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sine", detune: 0, gain: 1.0 },
    { type: "noise", detune: 0, gain: 0.5 },
  ],
  envelope: { attack: 0.001, decay: 0.2, sustain: 0.0, release: 0.3 },
  parameters: [
    {
      id: "punch",
      name: "Punch",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "tone",
      name: "Tone",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "room",
      name: "Room",
      type: "float",
      defaultValue: 0.2,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { punch: 0.7, tone: 0.5, room: 0.2, volume: 0.8 },
};

export default MbDrumsPlugin;
